google/ggrc-core
# GGRC-3776 Fix subquery for roles

# Issue description

MySQL returns an empty set of roles for the following query (using
only indexed field and ≥ 16 ids):

SELECT DISTINCT user_roles.role_id AS user_roles_role_id
FROM user_roles
WHERE user_roles.id IN (%s, %s, %s, %s, %s, %s, %s, %s,
                        %s, %s, %s, %s, %s, %s, %s, %s)

Prerequisites:
- UNIQUE KEY over (`role_id`,`context_id`,`person_id`)
- non-sequential ids in the parameters

mysql> explain SELECT DISTINCT user_roles.role_id AS user_roles_role_id  FROM user_roles  WHERE user_roles.id IN (5579, 3984, 311, 2295, 2134, 2103, 4701, 5011, 6104, 1369, 7245, 6727, 1400, 4313, 7713, 2462);
+----+-------------+------------+-------+------------------------------------+----------------------------+---------+------+------+---------------------------------------+
| id | select_type | table      | type  | possible_keys                      | key                        | key_len | ref  | rows | Extra                                 |
+----+-------------+------------+-------+------------------------------------+----------------------------+---------+------+------+---------------------------------------+
|  1 | SIMPLE      | user_roles | range | PRIMARY,unique_role_context_person | unique_role_context_person | 4       | NULL |   21 | Using where; Using index for group-by |
+----+-------------+------------+-------+------------------------------------+----------------------------+---------+------+------+---------------------------------------+


# Steps to test the changes

1. Restore the dump from August (Aug 1 and Aug 31 are reported to have this Audit, at the moment of writing the Audit with this id is already removed).
2. Navigate to /audits/203
3. Open People tab.

Expected result: 3 people with Program or Audit roles are displayed.
Actual result: "Uncaught TypeError: Cannot read property 'reify' of null" is thrown, only global roles are displayed in the tree view.

# Solution description

Mask the broken subquery with using a JOIN instead of it.

# Sanity checklist




# PR Review checklist
